### PR TITLE
Refactor FXIOS-6368 [v115] Refactor Notification service to not use tab queue

### PR DESCRIPTION
**Pull request by tarikeshaq**
[Jira ticket](https://mozilla-hub.atlassian.net/browse/FXIOS-6368)
[Github issue](https://github.com/mozilla-mobile/firefox-ios/issues/14304)

This also fixes https://github.com/mozilla-mobile/firefox-ios/issues/12678
[FXIOS-5424](https://mozilla-hub.atlassian.net/browse/FXIOS-5424) 

### Description
Needs https://github.com/mozilla/application-services/pull/5556 before it's taken out of draft

#### What's in this PR:
- The notification service no longer writes to the tab queue
- Notification service no longer modifies any state besides resetting a pref that tells the main process to poll for tabs
- Sent tabs should now show up quickly after a user opens the app (where previously it took 5 seconds)
- Deleted some redundant code that was an abstraction never used (`SyncDelegate`)

### Pull requests checks where applicable
- [x] Fill in the three TODOs above (tickets number and description of your work)
- [x] The PR name follows our [naming guidelines](https://github.com/mozilla-mobile/firefox-ios/wiki/Pull-Request-Naming-Guide)
- [x] Accessibility implemented and tested (minimum Dynamic Text and VoiceOver)
- [x] Unit tests written and passing
- [x] Documentation / comments for complex code and public methods

### CODE REVIEW

**Comment by tarikeshaq** [comment]
Still in draft, and I don't expect anyone to look at this just yet, but if anyone is curious please feel free to take a look!

I'll open it for review and rebase once the AS code is ready

**Comment by tarikeshaq** [comment]
Okay, the AS code is reviewed and ready to land, so this is ready for review.

The github ticket has context, but posted below for convenience too:
```
The way the notification service currently works is the following:

    Notification service receives the raw encrypted payload from push
    Notification service decrypts the payload
    Notification services passes the decrypted payload to an instance of an FxA client
    the FxA client converts the payload into an account event (lets assume it's a send tab)
    The notification service puts the tab in the tab queue
    The notification service then displays the tab as a notification
    When the app wakes up, it waits 5 seconds, then loads all tabs in the tab queue.

The above is mostly okay - however, we have a mechanism to retrieve missed send tabs by polling (in case the push notification never arrived) by passing the tabs in the queue and polling, we have two ways to get send tabs and it's possible for them to race - see https://github.com/mozilla-mobile/firefox-ios/issues/13136

Opening this ticket for myself to change the flow so that the Notification Service does not need the tab queue, and instead sets the polling pref to indicate that the app should hit the FxA server when it wakes up to get the tab.
```

**Comment by tarikeshaq** [comment]
To knowledge share the notification service and why it works the way it does, I'd love to write a short document - would https://github.com/mozilla-mobile/firefox-ios/wiki be the right place?

**Comment by tarikeshaq** [comment]
Also want to be explicit that although this is ready for review, it won't be green until the AS change is landed, this is deliberate in case there is any feedback here that requires changes there to minimize breaking changes - Ideally I would only land the AS change once this PR is approved then upgrade this PR to have the latest AS and land it

**Comment by OrlaM** [comment]
> To knowledge share the notification service and why it works the way it does, I'd love to write a short document - would https://github.com/mozilla-mobile/firefox-ios/wiki be the right place?

Yes, that would be great 🙏

File: Docs/notification-service.md
Comment (tarikeshaq) [inline Docs/notification-service.md:1]: I don't have permission to edit the wiki, and I wanted to make sure this document gets a quick review so I uploaded here.

I'll remove it once we add it to the wiki and before landing this PR

**Comment by tarikeshaq** [comment]
Thanks @OrlaM! Rendered document in https://github.com/mozilla-mobile/firefox-ios/blob/d0636a88be033d2867db5037e9b6fdb63e3ce078/Docs/notification-service.md

If we think it looks OK, we should move it to the wiki (I don't have permission) and I'll remove it from the PR

**Comment by lmarceau** [comment]
I think we'll be able to close https://mozilla-hub.atlassian.net/browse/FXIOS-5424 once that's merged?

**Comment by tarikeshaq** [comment]
> I think we'll be able to close [FXIOS-5424](https://mozilla-hub.atlassian.net/browse/FXIOS-5424) https://github.com/mozilla-mobile/firefox-ios/issues/12678 once that's merged?

Yes! I'll add that one to the PR description too 🌟

File: Providers/Profile.swift
Comment (tarikeshaq) [inline Providers/Profile.swift:725]: note to myself, we should probably set this to `0` instead, simpler and clearer

File: Docs/notification-service.md
Comment (lmarceau) [inline Docs/notification-service.md:10]:
```suggestion
To understand why the Notification Service and push work the way they do today, it's important to understand some of the constraints around how Firefox Accounts delivers push notifications.
```

File: Docs/notification-service.md
Comment (lmarceau) [inline Docs/notification-service.md:17]:
```suggestion
1. **Register with Autopush**, and receive a unique URL to pass to Firefox Accounts so Firefox Accounts can send push notifications to the applications.
```

File: Docs/notification-service.md
Comment (lmarceau) [inline Docs/notification-service.md:21]:
```suggestion
5. When Firefox Accounts sends a push notification to Autopush, it will encrypt the message using the public key it received from the application.
```

File: Docs/notification-service.md
Comment (lmarceau) [inline Docs/notification-service.md:35]:
```suggestion
Device commands don't carry their payload, instead, they carry an index to where the application can request their payload. For example, for send tab, the push notification (after it's decrypted by push) can look like the following:
```

File: Docs/notification-service.md
Comment (lmarceau) [inline Docs/notification-service.md:49]:
```suggestion
The application must then query the `url` to receive the content of the actual tab (there is a second layer of encryption there, but that's out of scope for this document).
```

**Review by lmarceau — COMMENTED** [review]
Some comments for the documentation, I'll review the rest soon!

File: Account/FxAPushMessageHandler.swift
Comment (lmarceau) [inline Account/FxAPushMessageHandler.swift:97]: Since we're refactoring this area, could we take the opportunity to remove deferred usage here?

File: Client/Application/AppDelegate+PushNotifications.swift
Comment (lmarceau) [inline Client/Application/AppDelegate+PushNotifications.swift:112]: nit
```suggestion
            profile.pollCommands(forcePoll: true)
```

File: Client/Application/AppLaunchUtil.swift
Comment (lmarceau) [inline Client/Application/AppLaunchUtil.swift:80]: Question about this to understand; if we have the `pollCommands` in `applicationDidBecomeActive`, it's necessary to have this one here because the `applicationDidBecomeActive` could be called before the RustFirefoxAccounts is startup?

File: Extensions/NotificationService/NotificationService.swift
Comment (lmarceau) [inline Extensions/NotificationService/NotificationService.swift:41]: So now we only have one event at a time? Just wanna understand the context of this change.

File: Extensions/NotificationService/NotificationService.swift
Comment (lmarceau) [inline Extensions/NotificationService/NotificationService.swift:204]: This string `SentTab_TabArrivingNotification_WithDevice_title` is not used anymore I think? If so we can remove from the `String.swift` file

File: Client/Application/AppDelegate+SyncSentTabs.swift
Comment (lmarceau) [inline Client/Application/AppDelegate+SyncSentTabs.swift:64]: I think `SentTabAction.TabSendCategory` is not used anymore and can be removed entirely?

File: Client/Application/AppDelegate+SyncSentTabs.swift
Comment (lmarceau) [inline Client/Application/AppDelegate+SyncSentTabs.swift:57]: Same thing for `TabSendURLKey` and `TabSendTitleKey`?

File: Extensions/NotificationService/NotificationService.swift
Comment (lmarceau) [inline Extensions/NotificationService/NotificationService.swift:215]: I think `SentTab_TabArrivingNotification_NoDevice_body` isn't used anymore as well, could be removed from `String.swift` ?

File: Extensions/NotificationService/NotificationService.swift
Comment (lmarceau) [inline Extensions/NotificationService/NotificationService.swift:217]: Same for `SentTab_TabArrivingNotification_WithDevice_body` string

File: Providers/Profile.swift
Comment (lmarceau) [inline Providers/Profile.swift:45]: nit, normally prefer comment format with `///`. This is also the provided comment with shortcut `option` + `cmd` + `/` when the cursor is on top of a function in Xcode.
```suggestion
/// This exists to pass in external context: e.g., the UIApplication can
/// expose notification functionality in this way.
```

File: Providers/Profile.swift
Comment (lmarceau) [inline Providers/Profile.swift:750]: nit, easier to read with guard IMO
```suggestion
                guard let commands = try? commands.get() else { return }
                let urls = commands.compactMap { command in
                    switch command {
                    case .tabReceived(_, let tabData):
                        let url = tabData.entries.last?.url ?? ""
                        return URL(string: url)
                    }
                }
                self.sendTabDelegate?.openSendTabs(for: urls)
```

File: Extensions/NotificationService/NotificationService.swift
Comment (tarikeshaq) [reply Extensions/NotificationService/NotificationService.swift:41]: Yes! and indeed some context here would have been helpful for the review,

In the past, the `handlePush` API could retrieve multiple events (in practice, multiple tabs in case some were accidentally lost)

However, that causes a bit of an unfortunate API. `handlePush` would get your missing tabs, put both them in the tab queue, but only show you the push notification for one.

 Logically that seems okay, but, because the notification service is running on an app extension it could race with the app when the app polls for commands, making it possible for a user to receive duplicate send tabs. To make easy to see the race:
- Notification service loads tabs into tab queue, and updates last-seen-tab-index
- Meanwhile, main process pollsCommands with the old last-seen-tab-index
- main process opens the tabs that is the result of pollCommands
- main process also opens the tabs in the tab queue, which duplicate the tabs it opened earlier



Now, the Notification service does not try to recover any missing tabs, and instead transforms the encrypted payload into a single notification to display to the user, then lets the app itself poll for commands to catch any missing tabs. It simplifies the job of the notification service to be: "Convert a push payload into something we want to show the user" and removes the side-effects

File: Client/Application/AppDelegate.swift
Comment (OrlaM) [inline Client/Application/AppDelegate.swift:145]: nit: You don't need to specify .default, it's a default parameter so you can remove the qos parameter altogether.

File: Docs/notification-service.md
Comment (OrlaM) [reply Docs/notification-service.md:1]: Added https://github.com/mozilla-mobile/firefox-ios/wiki/Notification-Service-and-Push

**Review by OrlaM — CHANGES_REQUESTED** [review]
Small nit, I will approve once Laurie's comments are addressed.

File: Client/Application/AppLaunchUtil.swift
Comment (tarikeshaq) [reply Client/Application/AppLaunchUtil.swift:80]: No this is a great catch! the `pollCommands` will only run the actual polling once the `RustFirefoxAccounts` has started up. i.e this was redundant - removing it!

File: Account/FxAPushMessageHandler.swift
Comment (tarikeshaq) [reply Account/FxAPushMessageHandler.swift:97]: Yes! Filed a follow up in https://github.com/mozilla-mobile/firefox-ios/issues/14518, I'll get started on that once I'm done with one

File: Client.xcodeproj/project.pbxproj
Comment (lmarceau) [inline Client.xcodeproj/project.pbxproj:18068]: Is that change expected?

**Review by lmarceau — APPROVED** [review]

File: Client/Experiments/Messaging/GleanPlumbMessageUtility.swift
Comment (tarikeshaq) [inline Client/Experiments/Messaging/GleanPlumbMessageUtility.swift:20]: cc @jhugman @adudenamedruby 

Encountered those breaking changes in the latest nightly. Propagated the new type where I could to get things to build and tests to pass but wanted to explicitly flag you

**Comment by tarikeshaq** [comment]
Bitrise is green! This should be good to land, I'll follow up with the deferred removal in another PR

**Pull request merged**
